src/NHibernate/Loader/Criteria/CriteriaQueryTranslator.cs

```diff
@@ -152,6 +154,14 @@ public QueryParameters GetQueryParameters()
 				{
 					lockModes[GetSQLAlias(subcriteria)] = lm;
 				}
+
+				if (subcriteria.HasRestrictions && subcriteria.JoinType == JoinType.LeftOuterJoin)
+				{
+					var collectionName = $"{rootCriteria.EntityOrClassName}.{subcriteria.Path}";
```

Comment:
This place looks hacky to me:
- We already have place where we collect all query collection persisters (see `CreateCriteriaCollectionPersisters`). I think it's a good place to collect this skip cache collection flags. And maybe collect them similarly to`criteriaCollectionPersisters' in `uncacheableCollectionPersisters`
- We should not rely on Left join anymore as now we support fetching collections with inner join too (with `SelectMode.Fetch`).

Reply:
It seems to me, `ShouldFetchCollectionPersister()` is exactly what we need to determine whether collection will be fetched or not. 
From other side, we can map `HasRestrictions` flags from subcriteries to assotiations, to determine if persister should be cached or not. 
That is why I have moved initialization of `uncacheableCollectionPersisters` to `CriteriaJoinWalker`. 
Is it ok for you?

Reply:
By "hacky" I meant we shouldn't use "magic strings" like `$"{rootCriteria.EntityOrClassName}.{subcriteria.Path}"`.  Will it work for multilevel fetch like `obj.Child.Collection` (it seems we need a test for this case, maybe it will work but it's not obvious to me)?

So instead I think it's better to collect directly all collection persisters `ICollectionPersister` with restrictions. That's why I suggested to look in to `CreateCriteriaCollectionPersisters` as that's the place where actual collection persisters are enumerated (though I'm not saying it's the only and best place to do the job)

To me this task looks like the following:
1) In translator (or walker) we need to find which collection persisters have restrictions;
2) And for loader we need to map "restricted" collection perstisters with persisters in `collectionPerstisters` and prepare `UncachedCollectionPersisters`

So to me it looks like we don't need to bother  "to determine whether collection will be fetched or not" - as long as we can obtain 'ICollectionPersister` for given criteria. And I don't see why `OuterJoinableAssociation` needs to be modified - it seems all information can be obtained from translator/walker.

Reply:
Ok, if I understood well, we shouldn't add collection into cache if:
1) There are restrictions and `subcriteria.JoinType == JoinType.LeftOuterJoin`
2) There are restrictions and `subcriteria.JoinType == JoinType.InnerJoin` and  `SelectMode == SelectMode.Fetch`, right?

And it wasn't clear for me how to get selection mode for subcriteria, but it seems that I can get it in this way - `rootCriteria.GetSelectMode(subcriteria.Path)`, it isn't?

Reply:
Could you have a look at my last commit, please? Is it right direction?

Reply:
Yep. But as I already said I think that logic that checks `Join` types and `SelectMode` is unnecessary here. No need to duplicate code - those checks will be executed by the routine that prepares collection persisters to fetch.
 
I've refactored and made a few clean ups to it (see my latest [commit](https://github.com/nhibernate/nhibernate-core/pull/460/commits/c25658e19988ea5178d82b82b8ef99c52fed5d3d)). Let me know if you disagree or see any issues.